File: python/headers/day3.py
# -*- coding: utf-8 -*-
"""
Day 3 of Advent of Code 2020

This script looks at moviing through a map counting the trees (#) as it does.
The psoition increases by an amount (grad) each time.

Tom Kite - 03/12/2020
"""

from aoc_tools.advent_timer import Advent_Timer


def readfile(name):
    with open(name) as file:
        data = [line.strip() for line in file]
    return data


def slide_down(pos, grad):
    return [pos[i] + grad[i] for i in range(len(pos))]


def check_tree(tree_map, pos):
    pos_x_eff = pos[0] % len(tree_map[0])
    return tree_map[pos[1]][pos_x_eff] == '#'


def full_slide(tree_map, grad, pos=[0, 0]):
    tree_count = 0
    while pos[1] < len(tree_map):
        if check_tree(tree_map, pos):
            tree_count += 1
        pos = slide_down(pos, grad)
    return tree_count


def part1(filename):
    data = readfile(filename)
    count = full_slide(data, [3, 1])
    print("On the way down you will hit {} trees.".format(count))


def part2(filename):
    data = readfile(filename)
    grads = [[1, 1], [3, 1], [5, 1], [7, 1], [1, 2]]
    mult_count = 1
    for grad in grads:
        mult_count *= full_slide(data, grad)
    print("The multiplication of trees encountered on each path is {}."
          .format(mult_count))


if __name__ == "__main__":
    timer = Advent_Timer()

    print("Part 1:")
    part1("../../data/day3.dat")
    timer.checkpoint_hit()

    print("\nPart 2:")
    part2("../../data/day3.dat")
    timer.checkpoint_hit()

    timer.end_hit()

File: C++/day3.cpp
/*
Advent of Code 2020 - Day 1

This code will search through a list of numbers and return the product of
the two or three that add to 2020

Tom Kite - 01/12/20
*/

#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;


vector<string> read_data(string filename)
{
    ifstream ifile;
    ifile.open(filename);
    if (!ifile) { throw; }

    vector<string> data{};
    string line;
    while (getline(ifile, line))
    {
        data.push_back(line);
    }
    ifile.close();
    return data;
}

class coord
{
public:
    int x,y;

    coord(int x_in, int y_in)
    {
        x = x_in;
        y = y_in;
    }

    coord operator+(coord other)
    {
        return coord(x + other.x, y + other.y);
    }
};

bool check_tree(vector<string> tree_map, coord position)
{
    int x_eff = position.x % tree_map[0].size();
    return tree_map[position.y][x_eff] == '#';
}

bool in_map(vector<string> tree_map, coord position)
{
    return position.y < tree_map.size();
}

int full_slide(vector<string> tree_map, coord grad, coord pos=coord{0,0})
{
    int tree_count{0};
    while (in_map(tree_map, pos))
    {
        if (check_tree(tree_map, pos)) tree_count++;
        pos = pos + grad;
    }
    return tree_count;
}

int main()
{
    vector<string> data;
    try { data = read_data("../data/day3.dat"); }
    catch (...) { cout << "Could not open file!" << endl; return 1; }

    long long int part_1{0};
    long long int part_2{1};

    part_1 = full_slide(data, coord{3, 1});

    vector<coord> grads;
    grads.push_back(coord{1,1});
    grads.push_back(coord{3,1});
    grads.push_back(coord{5,1});
    grads.push_back(coord{7,1});
    grads.push_back(coord{1,2});

    for (int i{0}; i<grads.size(); i++)
    {
        part_2 *= full_slide(data, grads[i]);
    }

    cout << part_1 << " " << part_2 << endl;

    return 0;
}
